File: api/Transaction.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package grakn.client.api;

import grakn.client.api.concept.ConceptManager;
import grakn.client.api.logic.LogicManager;
import grakn.client.api.query.QueryFuture;
import grakn.client.api.query.QueryManager;
import grakn.protocol.TransactionProto;

import java.util.stream.Stream;

public interface Transaction extends AutoCloseable {

    boolean isOpen();

    Type type();

    GraknOptions options();

    ConceptManager concepts();

    LogicManager logic();

    QueryManager query();

    void commit();

    void rollback();

    void close();

    enum Type {
        READ(0),
        WRITE(1);

        private final int id;
        private final boolean isWrite;

        Type(int id) {
            this.id = id;
            this.isWrite = id == 1;
        }

        public static Type of(int value) {
            for (Type t : values()) {
                if (t.id == value) return t;
            }
            return null;
        }

        public int id() {
            return id;
        }

        public boolean isRead() { return !isWrite; }

        public boolean isWrite() { return isWrite; }

        public TransactionProto.Transaction.Type proto() {
            return TransactionProto.Transaction.Type.forNumber(id);
        }
    }

    interface Extended extends Transaction {

        TransactionProto.Transaction.Res execute(TransactionProto.Transaction.Req.Builder request);

        QueryFuture<TransactionProto.Transaction.Res> query(TransactionProto.Transaction.Req.Builder request);

        Stream<TransactionProto.Transaction.ResPart> stream(TransactionProto.Transaction.Req.Builder request);
    }
}

File: src/connection/server/transaction.rs
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

use std::{fmt::Debug, time::Duration};

use futures::Stream;
use typedb_protocol::transaction as transaction_proto;

use crate::{
    common::{
        rpc::builder::transaction::{commit_req, open_req, rollback_req},
        Result, ServerRPC, TransactionRPC, TransactionType,
    },
    connection::core,
    query::QueryManager,
};

#[derive(Clone, Debug)]
pub struct Transaction {
    pub type_: TransactionType,
    pub options: core::Options,
    pub query: QueryManager,
    rpc: TransactionRPC,
}

impl Transaction {
    pub(crate) async fn new(
        session_id: &[u8],
        transaction_type: TransactionType,
        options: core::Options,
        network_latency: Duration,
        server_rpc: &ServerRPC,
    ) -> Result<Self> {
        let open_req = open_req(
            session_id.to_vec(),
            transaction_type.to_proto(),
            options.to_proto(),
            network_latency.as_millis() as i32,
        );
        let rpc = TransactionRPC::new(server_rpc, open_req).await?;
        Ok(Transaction { type_: transaction_type, options, query: QueryManager::new(&rpc), rpc })
    }

    pub async fn commit(&mut self) -> Result {
        self.single_rpc(commit_req()).await.map(|_| ())
    }

    pub async fn rollback(&mut self) -> Result {
        self.single_rpc(rollback_req()).await.map(|_| ())
    }

    pub(crate) async fn single_rpc(
        &mut self,
        req: transaction_proto::Req,
    ) -> Result<transaction_proto::Res> {
        self.rpc.single(req).await
    }

    pub(crate) fn streaming_rpc(
        &mut self,
        req: transaction_proto::Req,
    ) -> impl Stream<Item = Result<transaction_proto::ResPart>> {
        self.rpc.stream(req)
    }

    // TODO: refactor to delegate work to a background process
    pub async fn close(&self) {
        self.rpc.close().await;
    }
}
